lib/puppet/pops/types/string_converter.rb

```diff
@@ -0,0 +1,923 @@
+            buf << children_indentation.padding
+          elsif !(format.alt? && is_a_or_h?(v))
+            buf << ' '
+          end
+        end
+        buf << str_val
+      end
+      buf << delims[1]
+      buf
+    else
+      raise FormatError.new('Array', format.format, 'asp')
+    end
+  end
+
+  def is_a_or_h?(x)
+    x.is_a?(Array) || x.is_a?(Hash)
+  end
+
+  def is_container?(t)
+    case t
+    when PArrayType, PHashType, PStructType, PTupleType
+      true
+    else
+      false
+    end
+  end
+
+  # @api private
+  def string_PTupleType(val_type, val, format_map, indentation)
+    string_PArrayType(val_type, val, format_map, indentation)
+  end
+
+  # @api private
+  def string_PIteratorType(val_type, val, format_map, indentation)
+    v = val.to_a
+    _convert(@@type_calculator.infer_set(v), v, format_map, indentation)
+  end
+
+  # @api private
+  def string_PHashType(val_type, val, format_map, indentation)
+    format         = get_format(val_type, format_map)
+    sep            = format.separator  || DEFAULT_HASH_FORMAT.separator
+    assoc          = format.separator2 || DEFAULT_HASH_FORMAT.separator2
+    string_formats = format.container_string_formats || DEFAULT_CONTAINER_FORMATS
+    delims         = format.delimiter_pair(DEFAULT_HASH_DELIMITERS)
+
+    sep = format.alt? ? "#{sep}\n" : "#{sep} "
+
+    cond_break     = ''
+    padding        = ''
+
+    case format.format
+    when :a
+      # Convert to array and use array rules
+      array_hash = val.to_a
+      _convert(@@type_calculator.infer_set(array_hash), array_hash, format_map, indentation)
+
+    when :h, :s, :p
+      indentation = indentation.indenting(format.alt? || indentation.is_indenting?)
+      buf = ''
+      if indentation.breaks?
+        buf << "\n"
+        buf << indentation.padding
+      end
+
+      children_indentation = indentation.increase
+      if format.alt?
+        cond_break = "\n"
+        padding = children_indentation.padding
+      end
+      buf << delims[0]
+      buf << cond_break  # break after opening delimiter if pretty printing
+      buf << val.map do |k,v|
+        key_type = @@type_calculator.infer_set(k)
+        val_type = @@type_calculator.infer_set(v)
+        key = _convert(key_type, k, is_container?(key_type) ? format_map : string_formats, children_indentation)
+        val = _convert(val_type, v, is_container?(val_type) ? format_map : string_formats, children_indentation)
+        "#{padding}#{key}#{assoc}#{val}"
+      end.join(sep)
+      if format.alt?
+        buf << cond_break
+        buf << indentation.padding
+      end
+      buf << delims[1]
+      buf
+    else
+      raise FormatError.new('Hash', format.format, 'hasp')
+    end
+  end
+
+  # @api private
+  def string_PStructType(val_type, val, format_map, indentation)
+    string_PHashType(val_type, val, format_map, indentation)
+  end
+
+  # @api private
+  def string_PType(val_type, val, format_map, _)
+    f = get_format(val_type, format_map)
```

Comment:
This should be merged with the TypeFormatter. It's prepared for that and should utilize the array/hash formatting used here. Can be done as a separate JIRA.